pkg/resources/mutation/cluster/provisioningcluster.go

```diff
@@ -53,10 +97,258 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
 	if err != nil {
 		return nil, err
 	}
-	response := &admissionv1.AdmissionResponse{}
-	if err := mutation.SetCreatorIDAnnotation(request, response, request.Object, cluster.DeepCopy()); err != nil {
-		return nil, fmt.Errorf("failed to set creatorIDAnnotation: %w", err)
+
+	if request.Operation == admissionv1.Create {
+		// Set Annotation on the cluster
+		annotations := cluster.GetAnnotations()
+		if annotations == nil {
+			annotations = map[string]string{}
+		}
+		annotations[auth.CreatorIDAnn] = request.UserInfo.Username
+		cluster.SetAnnotations(annotations)
 	}
+
+	response, err := m.handlePSACT(request, cluster)
+	if err != nil {
+		return nil, err
+	}
+	if response.Result != nil {
+		return response, nil
+	}
+
 	response.Allowed = true
+	if err = patch.CreatePatch(request.Object.Raw, cluster, response); err != nil {
+		return nil, fmt.Errorf("failed to create patch: %w", err)
+	}
 	return response, nil
 }
+
+// handlePSACT updates the cluster and an underlying secret to support PSACT.
+// If a PSACT is set in the cluster, handlePSACT generates an admission configuration file, put it into a secret,
+// updates the cluster's spec to mount the secret to the control plane nodes, and configures kube-apisever to use the admission configuration file;
+// If a PSACT is unset in the cluster, handlePSACT does the cleanup for both the secret and cluster's spec.
+func (m *ProvisioningClusterMutator) handlePSACT(request *admission.Request, cluster *v1.Cluster) (*admissionv1.AdmissionResponse, error) {
+	if cluster.Name == "local" || cluster.Spec.RKEConfig == nil {
+		return admission.ResponseAllowed(), nil
+	}
+	parsedVersion, err := psa.GetClusterVersion(cluster.Spec.KubernetesVersion)
+	if err != nil {
+		return nil, fmt.Errorf("[provsioning cluster mutator] failed to parse cluster version: %w", err)
+	}
+	if parsedRangeLessThan123(parsedVersion) {
+		return admission.ResponseAllowed(), nil
+	}
+
+	secretName := fmt.Sprintf(SecretName, cluster.Name)
+	mountPath := fmt.Sprintf(MountPath, GetRuntime(cluster.Spec.KubernetesVersion))
+	templateName := cluster.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName
+
+	switch request.Operation {
+	case admissionv1.Delete:
+		err := m.secret.Delete(cluster.Namespace, secretName, &metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
What if the error is not nil but it's also not an `IsNotFound` API error? Couldn't the mutator fail to delete an existing secret? If so, I'd suggest updating this to the error handling pattern on line 171

Reply:
In this case, the user will see his operation is denied in the Rancher UI(the HTTP request fails), and he needs to check the rancher-webhook's logs if he wants to see the error message. 
I do not return the error to the user, like on line 171, because it is an internal error that should not be exposed to the user directly. As a compassion, on line 171, the error happens because the user sets a non-exist PSACT, in such case, we should return the error to the user to let him know. 
I am open to more discussion on this.

Reply:
There will not be a conflict even If the secret is left in the cluster because Rancher will either update the existing secret next time user updates the cluster or tries to delete the secret again next time user deletes the cluster.

Reply:
Ahh ok, thank you for clarifying. I think that is fine then